File: clojure/raindrops/src/raindrops.clj
(ns raindrops
  (:require [clojure.string :refer [join]]))

(defn test-for [num divisor word]
  (if (zero? (mod num divisor)) [word] []))

(defn convert [num]
  (let [check (partial test-for num)
        words (concat (check 3 "Pling")
                      (check 5 "Plang")
                      (check 7 "Plong"))]
    (if (empty? words) (str num) (join words))))

File: haskell/raindrops/src/Raindrops.hs
module Raindrops (convert) where

factors :: [(Int, String)]
factors = [(3, "Pling"), (5, "Plang"), (7, "Plong")]

convert :: Int -> String
convert n
  | null names = show n
  | otherwise  = names
  where names = factors >>= factorName
        factorName (f, name) = if n `mod` f == 0 then name else ""
